### PR TITLE
Feat/virtual joystick

### PR DESCRIPTION
**Pull request by jga-dev**
**New feature added: Basic Virtual Joystick**

Hi, I decide implement this feature because it appears to be in high demand. Hope you find interesting.

**Description:**
- Joystick can be activated by button on bottom (see picture)
- For resolutions less than 650px wide it's active by default
- When resolution ir less than 650px and chat is active, joystick is hidden because it interferes with the chat

Captura de Pantalla 2023-02-20 a las 1 20 20

For implement this, I use this component: https://github.com/elmarti/react-joystick-component

Note: This feature will conflict with this PR:  https://github.com/kevinshen56714/SkyOffice/pull/62
I can take care of resolving conflicts

Related issues: 
- https://github.com/kevinshen56714/SkyOffice/issues/23
- https://github.com/kevinshen56714/SkyOffice/issues/60

### CODE REVIEW

**Review by DaxChen — CHANGES_REQUESTED** [review]

**Comment by DaxChen** [comment]
This is awesome! Thanks for your contribution, please feel free to reach out if any of the comments are unclear or if you don't agree on any of them!

File: client/src/components/MobileVirtualJoystick.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import React, { useRef, useState, useEffect } from 'react'
+import styled from 'styled-components'
+import 'emoji-mart/css/emoji-mart.css'
```
[the remark this answers is not in the document]
Comment (jga-dev) [reply client/src/components/MobileVirtualJoystick.tsx:3]: Sure, you are right, it is not necessary.

File: client/src/components/Joystick.jsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
+import { Joystick } from 'react-joystick-component'
```
[the remark this answers is not in the document]
Comment (jga-dev) [reply client/src/components/Joystick.jsx:1]: No problem 👍  I'll change it in a moment

File: client/src/components/MobileVirtualJoystick.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  flex-direction: column;
+`
+
+const JoystickWrapper = styled.div`
+  margin-top: auto;
+  align-self: flex-end;
+`
+
+export default function MobileVirtualJoystick() {
+  const inputRef = useRef<HTMLInputElement>(null)
```
[the remark this answers is not in the document]
Comment (jga-dev) [reply client/src/components/MobileVirtualJoystick.tsx:43]: Yes, it stayed there after trying some things, but in the end it is not used.

**Comment by jga-dev** [comment]
@DaxChen Hi, thanks for the review. I just uploaded a new commit with the changes. Anything else let me know 😊👍

**Comment by MariaJesusGomez** [comment]
Oohh, Sounds great! As soon as this PR is resolved I'll pull the project with this feature.

File: client/src/components/Joystick.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import { Joystick } from 'react-joystick-component'
+
+export interface JoystickMovement {
+  isMoving: boolean
+  direction: Direction
+}
+
+type Direction = {
```
Comment (DaxChen) [inline client/src/components/Joystick.tsx:8]: let's use interface instead of type alias here

File: client/src/components/Joystick.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  left: boolean
+  right: boolean
+  up: boolean
+  down: boolean
+}
+
+type Props = {
```
Comment (DaxChen) [inline client/src/components/Joystick.tsx:15]: same, interface instead of type

File: client/src/components/Joystick.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  onDirectionChange: (arg: JoystickMovement) => void
+}
+
+const angleToDirections = (angle: number, out?: Partial<Direction> | boolean): Direction => {
+  const outObj: Direction = {
+    left: false,
+    right: false,
+    up: false,
+    down: false,
+    ...(typeof out === 'object' ? out : {}),
```
Comment (DaxChen) [inline client/src/components/Joystick.tsx:25]: this got me a little confused, if we're setting all left, right, up, down to false anyway, is there a reason to merge the input object here?
I would recommend getting rid of the second arg here and at the callsite.

File: client/src/components/MobileVirtualJoystick.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  useEffect(() => {}, [showJoystick, showChat])
+
+  const handleMovement = (movement: JoystickMovement) => {
+    game.myPlayer?.handleJoystickMovement(movement)
+  }
+
+  return (
+    <Backdrop>
+      <Wrapper>
+        {!(showChat && isSmallScreen) && showJoystick && (
```
Comment (DaxChen) [inline client/src/components/MobileVirtualJoystick.tsx:65]: `isSmallScreen` is a function but it doesn't seem like you're calling it, some typo here?

**Review by DaxChen — CHANGES_REQUESTED** [review]
Looking good! Thanks for the fast response!

File: client/src/components/Joystick.tsx (hunk 4 above)
Comment (jga-dev) [reply client/src/components/Joystick.tsx:8]: Ok 👍

File: client/src/components/Joystick.tsx (hunk 6 above)
Comment (jga-dev) [reply client/src/components/Joystick.tsx:25]: Sorry about this, my bad.. I was doing some tests and I forgot refactor it 😅 I'll change right now 👍 You'r right, only 1 argument it's needed

File: client/src/components/MobileVirtualJoystick.tsx (hunk 7 above)
Comment (jga-dev) [reply client/src/components/MobileVirtualJoystick.tsx:65]: Uff you'r right 😅, while I'm at it, I'll take the opportunity to make this cleaner

**Comment by jga-dev** [comment]
@DaxChen Hi, thanks again for your review. I've applied all the changes you have commented. If you see anything else feel free to let me know.

**Review by DaxChen — APPROVED** [review]
Awesome!! Thanks for cleaning up the window resize listener hook part, it's way cleaner now!

**Pull request merged**